widget-fulfillment/src/utilities.ts

```diff
@@ -0,0 +1,159 @@
+import { ArgOptional, Transaction, TransactionContext} from '@dbos-inc/dbos-sdk';
+import { Knex } from 'knex';
+
+type KnexTransactionContext = TransactionContext<Knex>;
+
+export enum OrderStatus {
+  PENDING = 0,
+  FULFILLED = 1,
+  PAID = 2,
+  ASSIGNED = 3,
+  CANCELLED = -1,
+}
+
+export interface Packer {
+  packer_name: string;
+  order_id: number | null;
+  expiration: Date | null;
+  timeLeft?: number;
+}
+
+export interface OrderPacker {
+  order_id: number;
+  order_status: OrderStatus;
+  product_id: number;
+  product: string;
+  packer_name: string | null;
+}
+
+export interface OrderWithProduct {
+  order_id: number;
+  order_status: OrderStatus;
+  last_update_time: Date;
+  product_id: number;
+  product: string;
+}
+
+export const PRODUCT_ID = 1;
+
+export class FulfillUtilities {
+  @Transaction({readOnly: true})
+  static async popDashboard(ctx: KnexTransactionContext) {
+    const orders = await ctx.client<OrderPacker>('order_packer').select().orderBy(['order_id']);
+    const packers = await ctx.client<Packer>('packer').select().orderBy(['packer_name']);
+    for (const o of orders) {
+      if (o.order_status === OrderStatus.PAID && o.packer_name) {
+        o.order_status = OrderStatus.ASSIGNED;
+      }
+    }
+    for (const p of packers) {
+      if (p.expiration) {
+        p.timeLeft = Math.round((p.expiration.getTime() - new Date().getTime())/1000);
+      }
+    }
+    return {orders, packers};
+  }  
+
+  @Transaction()
+  static async cleanStaff(ctx: KnexTransactionContext) {
+    await ctx.client<Packer>('packer').whereNull('order_id').delete();
+  }
+
+  @Transaction()
+  static async cleanOrders(ctx: KnexTransactionContext) {
+    await ctx.client<OrderPacker>('order_packer').where({order_status: OrderStatus.FULFILLED}).delete();
+  }
+
+  @Transaction()
+  static async addOrder(ctx: KnexTransactionContext, product: OrderWithProduct) {
+    await ctx.client<OrderPacker>('order_packer').insert({
+      order_id: product.order_id,
+      order_status: product.order_status,
+      product_id: product.product_id,
+      product: product.product,
+      packer_name: null,
+    }).onConflict(['order_id']).ignore();
+  }
+
+  @Transaction()
+  static async getUserAssignment(ctx: KnexTransactionContext, packer_name: string, expiration: Date, @ArgOptional more_time: boolean | undefined) {
+    let packers = await ctx.client<Packer>('packer').where({packer_name}).select();
+    let newAssignment = false;
+    if (!packers.length) {
+      await ctx.client<Packer>('packer').insert({packer_name, order_id: null, expiration: null});
+      packers = await ctx.client<Packer>('packer').where({packer_name}).select();
+    }
+    if (packers[0].order_id && more_time) {
```

Comment:
We always expect to have 0 or one packer here, right? This is more of a question - is it common to use the plural form in such cases?

Reply:
For me it's common to use the plural when it's an array... zero or more.